add: Rocket league umu

emrebicer opened the pull request:
After seeing the discussion at #151 and PR #193, I wanted to introduce Umu to Rocket League as well. Finally, I had some free time and made Rocket League work with Umu. 

However, there is still an issue that I could not find a solution to yet. If I enable umu, I can run bakkesmod alongside rocket-league, however once bakkesmod tries to inject, it fails and gives an error message that goes `Injection failed, please download vc_redist.x86.exe and restart your PC`. A Google search made me believe that the error message is implying I should have `Microsoft Visual C++ Redistributable` in the prefix, however even after trying to install `vcrun2019` with winetricks  (`umu-run winetricks vcrun2019`) or manually running the installer.exe (VC_redist.x86.exe or x64) in the prefix through umu, I still get the same error message bakkesmod.

I have been looking into possible solutions but could not get far. I wonder if anyone here has a suggestion to find a solution for that.

So **TLDR**; RL works with umu, and bakkesmod runs but does not inject. Any ideas on how can I fix that?

emrebicer committed:
update rocket-league: add umu

emrebicer committed:
add README for rocket-league

fufexan commented:
In my case the game doesn't seem to launch at all.

emrebicer committed:
use WINEPREFIX for umu as well

emrebicer commented:
I guessed perhaps I was not adding umu to PATH correctly, but I just removed umu from my config and RL still seems to work without issues, and looks like the path is set correctly.

 I added WINEPREFIX to umu scripts as well, maybe legendary was not directing to the correct game files. Though I am not confident that was the issue for you. 

it could be helpful if you shared with which attributes/values you call rocket-league in your config and what the output is when you start rocket-league from the terminal

fufexan commented:
Currently testing directly from the repo using this patch
```diff
diff --git a/pkgs/default.nix b/pkgs/default.nix
index 234bbf5..576a177 100644
--- a/pkgs/default.nix
+++ b/pkgs/default.nix
@@ -86,6 +86,7 @@
       rocket-league = pkgs.callPackage ./rocket-league {
         wine = config.packages.wine-tkg;
         inherit (config.packages) umu;
+        useUmu = true;
       };
 
       star-citizen = pkgs.callPackage ./star-citizen {
```
Output log is
```
nix-gaming on  rocket-league-umu [≡✓] via   
› nix run .#rocket-league
warning: Git tree '/home/mihai/Documents/code/git/nix-gaming' is dirty
do you want to allow configuration setting 'allowInsecure' to be set to 'true' (y/N)? 
do you want to permanently mark this value as untrusted (y/N)? 
warning: ignoring untrusted flake configuration setting 'allowInsecure'.
Pass '--accept-flake-config' to trust it
do you want to allow configuration setting 'extra-substituters' to be set to 'https://nix-gaming.cachix.org' (y/N)? 
do you want to permanently mark this value as untrusted (y/N)? 
warning: ignoring untrusted flake configuration setting 'extra-substituters'.
Pass '--accept-flake-config' to trust it
do you want to allow configuration setting 'extra-trusted-public-keys' to be set to 'nix-gaming.cachix.org-1:nbjlureqMbRAxR1gJ/f3hxemL9svXaZF/Ees8vCUUs4=' (y/N)? 
do you want to permanently mark this value as untrusted (y/N)? 
warning: ignoring untrusted flake configuration setting 'extra-trusted-public-keys'.
Pass '--accept-flake-config' to trust it
[Core] INFO: Trying to re-use existing login session...
[cli] INFO: Preparing download for "Rocket League®" (Sugar)...
[Core] INFO: Parsing game manifest...
[Core] INFO: Install path: /home/mihai/Games/rocket-league/rocketleague
[Core] INFO: Selected CDN: fastly-download.epicgames.com (https)
[DLM] INFO: Found 0 missing files.
[cli] INFO: Download size is 0, the game is either already up to date or has not changed. Exiting...
[cli] INFO: Logging in...
[Core] INFO: Trying to re-use existing login session...
[cli] INFO: Checking for updates...
[Core] INFO: Getting authentication token...
[cli] INFO: Launching Sugar...
gamemodeauto:                                                                                                                                                                                     
gamemodeauto: 
gamemodeauto: 

nix-gaming on  rocket-league-umu [≡✓] via   took 8s 
› gamemodeauto: 
steamrt is up to date
UMU-Proton is up to date
/nix/store/836by2rrl5xxwnnc4kbd9v21j9gglmv7-gvfs-1.54.2/lib/gvfs/libgvfscommon.so: undefined symbol: g_task_set_static_name
Failed to load module: /nix/store/836by2rrl5xxwnnc4kbd9v21j9gglmv7-gvfs-1.54.2/lib/gio/modules/libgioremote-volume-monitor.so
/nix/store/836by2rrl5xxwnnc4kbd9v21j9gglmv7-gvfs-1.54.2/lib/gvfs/libgvfscommon.so: undefined symbol: g_task_set_static_name
Failed to load module: /nix/store/836by2rrl5xxwnnc4kbd9v21j9gglmv7-gvfs-1.54.2/lib/gio/modules/libgvfsdbus.so
/nix/store/xj09bp6qdnf5aswgw9wpjcx2n9267ggp-dconf-0.40.0-lib/lib/gio/modules/libdconfsettings.so: undefined symbol: g_assertion_message_cmpint
Failed to load module: /nix/store/xj09bp6qdnf5aswgw9wpjcx2n9267ggp-dconf-0.40.0-lib/lib/gio/modules/libdconfsettings.so
gamemodeauto: dlopen failed - libgamemode.so: cannot open shared object file: No such file or directory
gamemodeauto: dlopen failed - libgamemode.so: cannot open shared object file: No such file or directory
gamemodeauto: dlopen failed - libgamemode.so: cannot open shared object file: No such file or directory
ERROR: ld.so: object '/tmp/pressure-vessel-libs-WVY9R2/${PLATFORM}/libgamemodeauto.so.0' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
ERROR: ld.so: object '/tmp/pressure-vessel-libs-WVY9R2/${PLATFORM}/libgamemodeauto.so.0' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
wine: using kernel write watches, use_kernel_writewatch 1.
gamemodeauto: dlopen failed - libgamemode.so: cannot open shared object file: No such file or directory
fsync: up and running.
gamemodeauto: dlopen failed - libgamemode.so: cannot open shared object file: No such file or directory
wine: using kernel write watches, use_kernel_writewatch 1.
gamemodeauto: dlopen failed - libgamemode.so: cannot open shared object file: No such file or directory
wine: using kernel write watches, use_kernel_writewatch 1.
gamemodeauto: dlopen failed - libgamemode.so: cannot open shared object file: No such file or directory
wine: using kernel write watches, use_kernel_writewatch 1.
gamemodeauto: dlopen failed - libgamemode.so: cannot open shared object file: No such file or directory
wine: using kernel write watches, use_kernel_writewatch 1.
gamemodeauto: dlopen failed - libgamemode.so: cannot open shared object file: No such file or directory
wine: using kernel write watches, use_kernel_writewatch 1.
gamemodeauto: dlopen failed - libgamemode.so: cannot open shared object file: No such file or directory
wine: using kernel write watches, use_kernel_writewatch 1.
gamemodeauto: dlopen failed - libgamemode.so: cannot open shared object file: No such file or directory
gamemodeauto: dlopen failed - libgamemode.so: cannot open shared object file: No such file or directory
wine: using kernel write watches, use_kernel_writewatch 1.
gamemodeauto: dlopen failed - libgamemode.so: cannot open shared object file: No such file or directory
wine: using kernel write watches, use_kernel_writewatch 1.
gamemodeauto: dlopen failed - libgamemode.so: cannot open shared object file: No such file or directory
wine: using kernel write watches, use_kernel_writewatch 1.
gamemodeauto: dlopen failed - libgamemode.so: cannot open shared object file: No such file or directory
wine: using kernel write watches, use_kernel_writewatch 1.
gamemodeauto: dlopen failed - libgamemode.so: cannot open shared object file: No such file or directory
wine: using kernel write watches, use_kernel_writewatch 1.
gamemodeauto: dlopen failed - libgamemode.so: cannot open shared object file: No such file or directory
gamemodeauto: dlopen failed - libgamemode.so: cannot open shared object file: No such file or directory
gamemodeauto: dlopen failed - libgamemode.so: cannot open shared object file: No such file or directory
gamemodeauto: dlopen failed - libgamemode.so: cannot open shared object file: No such file or directory
gamemodeauto: dlopen failed - libgamemode.so: cannot open shared object file: No such file or directory
dbus[435206]: arguments to dbus_pending_call_block() were incorrect, assertion "pending != NULL" failed in file dbus-pending-call.c line 766.
This is normally a bug in some application using the D-Bus library.

  D-Bus not built with -rdynamic so unable to print a backtrace
  ```

emrebicer committed:
fix gamemode path

emrebicer commented:
I just noticed we were not adding gamemode to the path correctly, but I don't think that was the issue with your case anyway. Honestly, when I run rocket-league, the output is pretty similar, I don't see any noticeable difference, except the game runs normally. Perhaps we can try looking into umu logs like that;
```bash
$ UMU_LOG=debug rocket-league
```

fufexan commented:
Sure, here's the log.
[rl.log](https://github.com/user-attachments/files/16604206/rl.log)

emrebicer commented:
I just compared the output with the output on my end, it looks pretty similar except for the `wine: using kernel write watches, use_kernel_writewatch 1.` statements toward the end, where I don't have such logs. I don't think that tells anything helpful though. 

To debug further I would recommend creating a shell with umu-run and legendary-gl, they try to manually run the game directly through umu-run,
```
export GAMEID=umu-252950
export STORE=egs
export PROTON_VERB=runinprefix

legendary update Sugar --base-path ${location}
legendary launch Sugar --no-wine --wrapper "umu-run" --base-path ${location} --dry-run
```
With the dry run option, the last command should print the command legendary will run and just exit. Then you can try manually running this command without including gamemode or any other dependencies in the way. Also, it might be worth trying to setup the envvar PROTONPATH=GE-Proton to use a different version of proton.

But at this point, it becomes a little too much work and I am not confident that this kind of debugging can get us somewhere, so one option is to leave the PR in this state as a draft and come back to it when there are new releases or improvements from umu project. Honestly, it is a bit sad to see that even nix struggles with the "it works on my machine" problem.

fufexan commented:
@emrebicer following your instructions I managed to get it to launch after a `nix flake update`.

emrebicer commented:
That is very good to hear, so does the PR also work after the nix flake update by any chance? Or just the dry-run? If it is only the dry run, then perhaps you can try prepending the `gamemoderun` command before the dry-run output, so we can see if gamemode is somehow at fault

fufexan commented:
Only the dry-run works. Strangely enough, it also works if I prepend `gamemoderun`, but not when I do `nix run .#rocket-league`.

emrebicer commented:
I think I found the culprit here. I moved to another (physical) place where I don't have access to the desktop I have developed the changes for this PR, but I have my Nvidia desktop that is mentioned at #186 .

On this desktop. I can reproduce the issue you are having, I simply can't run Rocket League with Umu with this PR. I did do some digging and I believe the issue is with the environment variable `export PROTON_VERB=runinprefix`. By default, if this environment variable is not found, umu will set the option to `PROTON_VERB=waitforexitandrun` and I think this is the recommended way to use umu, however since I wanted to introduce bakkesmod option from nix-gaming I need to be able to run 2 programs at the same time using umu, so I set this env var to `runinprefix` as it is recommended in the [umu project wiki](https://github.com/Open-Wine-Components/umu-launcher/wiki/Frequently-asked-questions-(FAQ)).

Again, it was working without issues on my other desktop, but apparently, there is something wrong with it so it doesn't work on my Nvidia machine or your environment...

So I suggest for now we focus on making Rocket League work in any environment with umu and ignore bakkesmod case at the moment. Given that, I pushed a commit that removes the line where we set the mentioned env var. With that change, Rocket League Umu works on my Nvidia machine as well without issues. (A side note: there are no performance issues with umu, so this PR might lead to closing #186  as well.)

Can you please try with the latest change as well? If it also works for you without issues, then we can focus on how we can support bakkesmod with Umu as well, I guess I can open an issue about it at the Umu project repo and ask for their help/recommendations about it.

fufexan commented:
@emrebicer I'm terribly sorry for the delay, the notification must have slipped my view.

I've just tested this, seems to work well despite the gamemode errors (which are probably a skill issue on my part).

emrebicer commented:
No worries, I also get many gamemode errors that go `gamemodeauto: dlopen failed - libgamemode.so: cannot open shared object file: No such file or directory`. I don't know if this is the same issue, but similar issues are still open at upstream like [here](https://github.com/FeralInteractive/gamemode/issues/254) and [there](https://github.com/Open-Wine-Components/umu-launcher/issues/81).

However despite the error, looks like gamemode is being activated, as the output of `gamemoded -s` is returning `gamemode is active` when Rocket League is running. So I am not sure if this is something we can fix / should be worried about.

fufexan commented:
Okay, then hopefully it will be fixed upstream. Thanks for looking into it.

fufexan commented:
I'd suggest trying older `vcrun20xx` as well, and if that doesn't work, try to copy the resulting dll to bakkesmod's directory along with creating a dll override. Not sure it will work, but it's worth a try.

emrebicer commented:
I tried installing all the vcrun20xx versions that I see on the [winetricks verbs](https://github.com/Winetricks/winetricks/blob/master/files/verbs/all.txt), but none of them resulted in a different behavior. I am not sure about the resulting DLLs, but I found vcruntime140.dll and kernel32.dll copied them to the same folder as bakkesmod.exe and tried rerunning things, it did not change any behavior either. 

I don't know how to override DLLs on wine, googling suggested using winecfg, not sure how well winecfg is supported for proton but gave it a shot and ran it by setting the target umu prefix. Under the libraries tab, there are some DLLs but I did not see a way to override the DLL by path, it lets me choose the load order for native or builtin. So if you have some, I would appreciate some tips for achieving the overriding DLL suggestion.

Other than that I found other things such as, if I run `umu-run winetricks <any_verb>` before running RL, afterward it actually lets me run RL using `PROTON_VERB=runinprefix` env var. However, the result is not so satisfying as there is some flickering happening and performance is noticeably worse.

Another thing I tried was to create a .bat script to launch bakkesmod and RL at the same time just for testing purposes (I wanted to create an environment where I run both of them without setting `PROTON_VERB=runinprefix` while using umu). It kinda worked, they both started and bakkesmod injected without throwing the missing dependency error... However, then I realized if I start them this way, RL won't respond to keyboard/mouse and screen resize events and just be stuck at the entrance screen to the game (it is not frozen but I could not just proceed to play the game). My best guess is that there might be a proton limitation somehow that prevents focusing on multiple windows. However, the important finding here is that bakkesmod injects fine without throwing an error if it is started without the `PROTON_VERB=runinprefix`  env var.

So my take on these findings is; for whatever reason using `PROTON_VERB=runinprefix` messes with the behaviour of the programs, and AFAIK this is the only way to run multiple programs at once. I already opened an issue regarding that on the umu repository [here](https://github.com/Open-Wine-Components/umu-launcher/issues/194). However the issue is kind of waiting for more reports atm and their reasonable take is "This is technically not a bug with the launcher, but rather a possible constraint with an upstream component or an OS issue.".

fufexan committed:
rocket-league: use GE-Proton when using umu

fixup some nits

emrebicer committed:
Fix Rocket League EAC option

EAC option is enabled by default, and can be disabled
by passing in -noeac to the legeandary launch command.

emrebicer committed:
Remove wine option from Rocket League, force umu

With the introduction of easy anticheat to RL, the wine
option does not work anymore, so remove the wine option
and force umu launcher.

emrebicer committed:
Add umuProtonPath option

This is the environment variable that dictates which
proton should the umu-launcher use. It defaults to
GE-Proton but can be customized to other proton builds.

emrebicer committed:
Update Rocket League readme

fufexan commented on pkgs/rocket-league/bakkesmod.nix:
```diff
@@ -16,7 +17,7 @@ let
   };
 
-  bakkesmodExePath = "${location}/drive_c/Program Files/BakkesMod/BakkesMod.exe";
+  bakkesmodExePath = "$HOME/Games/umu/umu-252950/drive_c/Program Files/BakkesMod/BakkesMod.exe";
```
Why not change the `location` itself?

emrebicer replied:
Updated it to use the location parameter, rather than hardcoding the wineprefix path

fufexan commented on pkgs/rocket-league/default.nix:
```diff
@@ -6,23 +6,14 @@
+  location ? "$HOME/Games/umu/umu-252950",
   dxvk_hud ? "compiler",
   callPackage,
   enableEAC ? true,
```
Since `eac-runtime` was removed, we could add a warning that this flag is no longer used.

emrebicer replied:
I figured out the eac-runtime is not used at all with rocket league, maybe the way it was integrated/bundled was different compared to other games. But we don't need to pass the eac-runtime to the game in order to launch with eac. Simply running it through legendary enables and runs eac. 

However `enableEAC` is still used, if the flag is set to false, `-noeac` parameter is passed to the game, which then launches the game without EAC.

fufexan replied:
Makes sense, thanks for the explanation.

fufexan commented on pkgs/rocket-league/README.md:
```diff
@@ -0,0 +1,34 @@
+    home.packages = with pkgs; [
+      (inputs.nix-gaming.packages.${pkgs.system}.rocket-league.override {
+        enableBakkesmod = true;
+        enableEAC = false;
```
Isn't this a no-op now?

emrebicer replied:
`enableEAC` is still useful (see comment above), but the `enableBakkesmod` flag eventhough would allow installing and running bakkesmod, it won't work with the game yet. For that option we can "add a warning that this flag is no longer used", but tbh I have no idea how we can do that. I already tried to mention that the bakkesmod won't work yet through nix-gaming to the readme, but is that enough?

fufexan replied:
We could do something like this before calling `stdenv.mkDerivation`.
```nix
assert lib.assertMsg enableBakkesmod "The option `enableBakkesmod` is set to `true`. BakkesMod does not work through umu yet. See https://github.com/Open-Wine-Components/umu-launcher/issues/283.";
```

emrebicer replied:
Added the suggested assertion. When enableBakkesmod is set to true, it fails with an error, and `nixos-rebuild` fails, otherwise builds okay and can play RL

emrebicer replied:
I now noticed that after my most recent commit the CI job `NIXPKGS_ALLOW_INSECURE=1 nix flake check --impure` seems to fail, it does not fail locally so I am a bit confused, the error mentions `while checking the derivation 'packages.x86_64-linux.dxvk'` which I did not edit but did I do something wrong?

fufexan replied:
I'm gonna rerun, maybe it was a hiccup.

fufexan replied:
A rebase should fix it.

emrebicer replied:
Yep, seems good now, only the wine-osu CI stage seem to fail, but that is the same on master

fufexan replied:
Yeah, I haven't been able to get it working so I let it be. I might try to fix it sometime.

fufexan reviewed (commented):
Runs fine, overall lgtm.

emrebicer committed:
Use location variable rather than hardcoding WINEPREFIX path

emrebicer committed:
Add assertion for Bakkesmod

Bakkesmod does not work with umu yet, so show an error
message when enabled.

emrebicer committed:
Merge branch 'fufexan:master' into rocket-league-umu

fufexan reviewed (approved):
Great work, LGTM!

fufexan commented:
@emrebicer wanted to let you know I got bakkesmod working again on master 1f64f73.

emrebicer commented:
@fufexan Seems to work as intended on my end; well done! Feel free to ping me about future requests for RL.
